ffc/main.py

```diff
@@ -1,51 +1,69 @@
+Parse command-line arguments and generate code from input UFL form files.
 """
 
+import argparse
 import cProfile
-import getopt
 import logging
-import os
+import pathlib
 import re
 import string
-import sys
 
 import ufl
 from ffc import __version__ as FFC_VERSION
-from ffc.backends import ufc
-from ffc import compiler
-from ffc import formatting
+from ffc import compiler, formatting
 from ffc.parameters import default_parameters
 
 logger = logging.getLogger(__name__)
 
-
-def info_version():
-    """Print version number."""
-    print("""\
-This is FFC, the FEniCS Form Compiler, version {0}.
-UFC backend version {1}, signature {2}.
-For further information, visit https://bitbucket.org/fenics-project/ffc/.
-
-Python {3} on {4}
-""".format(FFC_VERSION, ufc.__version__, ufc.get_signature(), sys.version, sys.platform))
-
-
-def info_usage():
-    """Print usage information."""
-    info_version()
-    print("""Usage: ffc [OPTION]... input.form
-
-For information about the FFC command-line interface, refer to
-the FFC man page which may invoked by 'man ffc' (if installed).
-""")
+# Specify command line options
+parser = argparse.ArgumentParser(
+    description="FEniCS Form Compiler (FFC, https://fenicsproject.org)")
+parser.add_argument(
+    "-l",
+    "--language",
+    type=str,
+    action='store',
+    choices=["ufc", "dolfin"],
+    default="ufc",
+    help="target language/wrappers")
+parser.add_argument(
+    "--version",
+    action='version',
+    version="%(prog)s " + ("(version {})".format(FFC_VERSION)))
+parser.add_argument("-d", "--debug", action='store_true', default=False,
+                    help="enable debug output")
+parser.add_argument("-v", "--verbose", action='store_true', default=False,
+                    help="verbose output")
+parser.add_argument("-p", "--profile", action='store_true', default=False,
+                    help="enable profiling")
+parser.add_argument("-o", "--output-directory", type=str, help="output directory")
+parser.add_argument(
+    "-r",
+    "--representation",
+    type=str,
+    action='store',
+    choices=('uflacs', 'tsfc'),
+    default="uflacs",
+    help="representation ")
+parser.add_argument("-q", "--quadrature-rule", default="auto",
+                    help="quadrature rule to apply")
+parser.add_argument("--quadrature-degree", default="auto",
+                    help="quadrature degree to apply")
+parser.add_argument('-f', action="append", default=[], dest="f", metavar="parameter=value",
+                    help="options passed through to parameter system")
+parser.add_argument("ufl_file", nargs='+', help="UFL file(s) to be read from")
+# parser.add_argument("--no-evaluate-basis-derivatives", action='store_true',
+#                     help="disable generation of code for evaluating basis derivatives")
+# parser.add_argument("-O", "--optimize", action='store_true', default=False,
+#                    help="apply optimizations during code generation")
```

Comment:
It's not clear that `--optimize` does anything at the moment (other than make following the FFC code confusing), and `--no-evaluate-basis-derivatives` did do nothing.

A generic option like `--optimize` isn't helpful. FFC should just optimize where it can. Any  optimization that is very expensive to perform (e.g., in the old quadrature code) or which could degrade performance should have a descriptive name and help string.

Reply:
Optimize indeed does a lot in UFLACS. The option enters UFLACS [here](https://github.com/FEniCS/ffcx/blob/99f2648663954a7a93d3768a7af05906f2a28e48/ffc/uflacs/build_uflacs_ir.py#L173). This at the same time gives the list of options which you could likely be moved closer to the user interface as you suggest.